libraries/standard/mqtt/src/mqtt.c

```diff
@@ -301,6 +301,231 @@ static MQTTStatus_t validatePublishParams( const MQTTContext_t * pContext,
+ * @param[in] filterIndex Index of the topic filter being examined.
+ *
+ * @return Returns whether the topic filter and the topic name match.
+ */
+static bool matchWildcardsSpecialCases( const char * pTopicFilter,
+                                        uint16_t topicFilterLength,
+                                        uint16_t topicNameLength,
+                                        uint16_t nameIndex,
+                                        uint16_t filterIndex );
+
+/**
+ * @brief Attempt to match topic name with a topic filter starting with a wildcard.
+ *
+ * If the topic filter starts with a '+' (single-level) wildcard, the function
+ * advances the @a pNameIndex by a level in the topic name.
+ * If the topic filter starts with a '#' (multi-level) wildcard, the function
+ * concludes that both the topic name and topic filter match.
+ *
+ * @param[in] pTopicFilter The topic filter containing the wildcard.
+ * @param[in] topicFilterLength Length of the topic filter.
+ * @param[in] pTopicName The topic name to check.
+ * @param[in] topicNameLength Length of the topic name.
+ * @param[in] filterIndex Index of the wildcard in the topic filter.
+ * @param[in,out] pNameIndex Index of character in topic name. This variable is
+ * advanced for `+` wildcards.
+ * @param[out] pMatch Whether the topic filter and topic name match.
+ *
+ * @return `true` if the caller of this function should exit; `false` if the caller
+ * should continue parsing the topics.
+ */
+static bool matchWildcards( const char * pTopicFilter,
+                            uint16_t topicFilterLength,
+                            const char * pTopicName,
+                            uint16_t topicNameLength,
+                            uint16_t filterIndex,
+                            uint16_t * pNameIndex,
+                            bool * pMatch );
+
+/**
+ * @brief Match a topic name and topic filter allowing the use of wildcards.
+ *
+ * @param[in] pTopicName The topic name to check.
+ * @param[in] topicNameLength Length of the topic name.
+ * @param[in] pTopicFilter The topic filter to check.
+ * @param[in] topicFilterLength Length of topic filter.
+ *
+ * @return `true` if the topic name and topic filter match; `false` otherwise.
+ */
+static bool matchTopicFilter( const char * pTopicName,
+                              uint16_t topicNameLength,
+                              const char * pTopicFilter,
+                              uint16_t topicFilterLength );
+
+/*-----------------------------------------------------------*/
+
+static bool matchWildcardsSpecialCases( const char * pTopicFilter,
+                                        uint16_t topicFilterLength,
+                                        uint16_t topicNameLength,
+                                        uint16_t nameIndex,
+                                        uint16_t filterIndex )
+{
+    /* The smallest topic filter accepted by this function is "/+". */
+    assert( topicFilterLength > 1 );
```

Comment:
Are we no longer asserting on NULL pointers?

Reply:
Also, do we even check this before calling this function? For example, with the function call of a topic name "a/" and topic filter "a":
```
MQTT_MatchTopic( "a/", 2, "a", 1, &match );
```
Unless I'm missing something, this will on line 2246 call 
```
matchTopicFilter( "a/", 2, "a", 1 );
```
which satisfies the conditions on line 480 and 484: 
```
while( ( 0 < 2 ) && ( 0 < 1 ) )
```
and 
```
if( pTopicName[ 0 ] == pTopicFilter[ 0 ] )
```
This will then call `matchWildcardsSpecialCases( "a", 1, 2, 0, 0 );`, triggering the assert since `topicFilterLength` = 1.

Reply:
> Are we no longer asserting on NULL pointers?

Thanks for catching. Have added the NULL-check asserts

Reply:
> Also, do we even check this before calling this function? For example, with the function call of a topic name "a/" and topic filter "a":
> 
> ```
> MQTT_MatchTopic( "a/", 2, "a", 1, &match );
> ```
> 
> Unless I'm missing something, this will on line 2246 call
> 
> ```
> matchTopicFilter( "a/", 2, "a", 1 );
> ```
> 
> which satisfies the conditions on line 480 and 484:
> 
> ```
> while( ( 0 < 2 ) && ( 0 < 1 ) )
> ```
> 
> and
> 
> ```
> if( pTopicName[ 0 ] == pTopicFilter[ 0 ] )
> ```
> 
> This will then call `matchWildcardsSpecialCases( "a", 1, 2, 0, 0 );`, triggering the assert since `topicFilterLength` = 1.

Have removed the assert check for the topic filter length as there exists an if-check in the logic that covers the "/+" case
`if( ( nameIndex == ( topicNameLength - 1U ) ) && ( filterIndex == ( topicFilterLength - 2U ) ) )`